### ISSUE TITLE
Clarify USM allocation accessibility by other devices

### ISSUE THREAD

**Issue by AerialMantis** [KhronosGroup/SYCL-Docs#186]
In §4.8.2 where the different kinds of USM are described table 99 the characteristics of the different kinds of USM allocation. Part of the table describes whether a USM allocation is accessible by another device. This is described for device and shared allocations as `Optional`, which I interpret to mean that it's implementation defined whether that is supported. However, for device allocations it also has `P2P` in brackets, while for shared allocations it doesn't, though the term `P2P` or `peer-to-peer` doesn't seem to be referenced anywhere else in the specification.

I was hoping to clarify in which circumstances a device or shared memory allocation can be expected to be accessible by other devices, i.e. is there a way to query this, and what is meant by `P2P` and what how this differs between device and shared allocations.

cc @jbrodman

**Comment by AerialMantis** [comment]
Following up on the discussion from the call, I thought I'd describe my understanding of the clarifications and actions we took from this.

So my understanding of the "Accessible By" column in table 99 is that it describes the capability of a USM allocation of a particular type to be accessed directly by another device other than it's initial location. Therefore the reason for the distinction between a device allocation being accessible by another device being considered "Optional (P2P)" and a shared allocation being accessible by another device being considered "Optional" is that both are optional since this capability depends on the particular combination of devices, and for a device allocation typically the memory will be allocated for a specific device, yet, some backends will provide a way to enable direct access by peer devices, i.e. P2P, whereas for a shared allocation it is part of the semantics that the allocation should be accessible by all devices capable of sharing that allocation.

Saying that we also discussed that we should clarify the term "P2P" and more generally the ways in which different USM allocations could provide access to other devices, even if as a non-normative note, and perhaps to provide some kind of query to express these capabilities so applications could reason about the behaviour.

**Comment by fraggamuffin** [comment]
Thank you.

**Comment by VerenaBeckham** [comment]
I'd like to bring this up again in a WG discussion.

**Comment by VerenaBeckham** [comment]
I would also add that the term "Migratable To" is not very clear and not mentioned anywhere else in the spec.

**Comment by TApplencourt** [comment]
With pleasure! Will put that on the Agenda : )  
I really dislike this section, is so confusing (migratable, and P2P are indeed used but never defined...)

**Comment by VerenaBeckham** [comment]
So, I suggest we
* remove the "Migratable To" columns
* optionally add a paragraph on "intention", but do we want to talk about how we intend the implementers to implement the different allocation types, or what relative performance a user can typically expect? Personally, I think there is more detailed description underneath this table which makes it sufficiently clear that device is on the device, host is on the host and shared could potentially migrate implicitly.

Additionally, we have two options:
1. replace the "Optional" and "Optional (P2P)" by "Implementation-defined"
2. replace the "Optional" and "Optional (P2P)" with "No" and anyone who wants to support this needs to create an extension

@gmlueck does Intel support one or the other of these without an extension at the moment? Then I guess it wouldn't be possible to restrict this capability in hind-sight and we would need to go for Option 1.

**Comment by VerenaBeckham** [comment]
I think the behaviour / documentation of `memcpy` is related to this:

> void memcpy(void* dest, const void* src, std::size_t numBytes) - Copies numBytes of data from the pointer src to the pointer dest. The dest and src parameters must each either be a host pointer or a pointer within a USM allocation that is accessible on the handler’s device.

If we memcpy between two devices then I presume "accessible" refers to the same concept as the "Optional" entries in the table?

**Comment by TApplencourt** [comment]
I don't think so. I think `memcpy` is allways possible between 2 device on the same context (because it's trivial to generate a `tmp-host-buffer`) Accessible mean same context here?

But you cannot necessarily "dereference" a device memory pointer from another device inside a kernel.

Aka
```
C{D1,D2};
Q1{C,D1}; Q2{C,D2};

auto P1 = malloc_device(N, Q1);
auto P2 = malloc_device(N, Q2);

Q1.memcopy(P1, P2) ; // Always ok?
Q1.parrallel_for(N, [](auto id){ P1[id] = P2[id] } ); // May or may not work ?
```

But maybe (more probably) I'm wrong...

**Comment by VerenaBeckham** [comment]
Later it also says that both allocations have to be in the same context:

> If a pointer is to a USM allocation, that allocation must have been created from the same context as the handler’s queue.

So, it seems unlikely that "accessible" == "same context", but that you require "accessible" && "same context".
It is not clear to me whether `memcpy` is allowed to use a `tmp-host-buffer`, or whether it is intended to be a straight copy. That's also something we should clarify.

**Comment by TApplencourt** [comment]
For memcopy, as user can always do:
```
std::vector tem_buffer(N);
e = Q1.memcopy(P1,N);
Q2.memcopy(N,P2,e);
```

So I was thinking that `memcopy(P1,P2);` can always made to "just-work" (this allocate host-memory but I guess that ok? I don't knoew)

Concerning accessible, Opencl said

>      Context
>    The environment within which the kernels execute and the domain in which synchronization and memory management is defined. The context includes a set of devices, the memory accessible to those devices, the corresponding memory properties and one or more command-queues used to schedule execution of a kernel(s) or operations on memory objects.

(https://registry.khronos.org/OpenCL/specs/unified/html/OpenCL_API.html) 

`the memory accessible to those devices` but maybe not "between device"? 

The language for SVM: here https://registry.khronos.org/OpenCL/specs/unified/html/OpenCL_API.html#_svm_sharing_granularity_coarse_and_fine_grained_sharing  but they don't seem to discuss any `device <-> device`.

I guess for use, we should clarify what we mean by "accessible", and what do we want for `memcopy`?

**Comment by gmlueck** [comment]
Catching up after being away for a few days ...

Personally, I think we could just remove [Table 69](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.USM.allocation.characteristics).  I think it adds confusion without really clarifying anything.

The "migratable to" columns seem like an implementation detail to me.  They describe one possible way that USM might be implemented.  For example, in a native CPU implementation, I don't think the memory buffers would really "migrate".  Instead, it might make sense to have a non-normative note explaining the intent of the three types of USM allocations and their expected performance characteristics.  (E.g. host USM stays resident on the host while shared USM may migrate between host and device(s).)

I think it was a mistake for the core SYCL spec to list "P2P" at all.  There is no such feature in the core spec.  As a result, I think that applications which strictly follow the core spec cannot assume that USM device memory is accessible from any device other than the device where it was allocated.  Some vendors have provided extensions that *do* allow this (e.g. Intel's [sycl_ext_oneapi_peer_access](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_peer_access.asciidoc)), however that is not a reason to list this feature as "optional" in the core spec.  In general, vendors are allowed to add extensions that extend the core spec, and there is no reason for the core spec to make note of these extensions.

(As an aside, we should consider making a KHR extension for P2P, but that is a separate discussion.)

It would probably be useful if the spec was more clear about when a USM buffer is accessible.  I think our intent was:

* A USM host buffer is accessible from the host and from any device that is in the same context as the USM allocation.
* A USM shared buffer is accessible from the host and from any device that is in the same context as the USM allocation.
* A USM device buffer is accessible only from the device specified in the USM allocation.

There are additional rules that govern when it is safe for two kernels to concurrently access the same USM buffer, when it is safe for the host and a kernel to concurrently access the same USM buffer, and when it is safe to use atomic operations in these situations.  It is not clear to me what our intent was for all the edge cases in these scenarios, so I would suggest tackling this part separately.

**Comment by gmlueck** [comment]
Regarding the explicit memory copy APIs, and this comment from @TApplencourt:

> So I was thinking that `memcopy(P1,P2);` can always made to "just-work" (this allocate host-memory but I guess that ok? I don't knoew)

I think this *is* clear in the spec.  All of the explicit memory copy APIs have wording like this:

> The `dest` and `src` parameters must each either be a host pointer or a pointer within a USM allocation that is accessible on the handler’s device. If a pointer is to a USM allocation, that allocation must have been created from the same context as the handler’s queue.

Therefore, passing arbitrary `P1` and `P2` is not guaranteed to "just work".

**Comment by TApplencourt** [comment]
cuda have`cudaMemcpyPeer`  (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html), where cudaMemcpy is only for host-device

**Comment by TApplencourt** [comment]
Maybe for new bullet points to clarify shared aspects. Something like:
```
- A USM host buffer is accessible from the host and from any device that is in the same context as the USM allocation.
- A USM shared buffer is accessible from the host  the device specificed USM allocation if they support `shared_non_atomic` aspect.
- A USM shared buffer is accessible from the host and from any device that is in the same context as the USM allocation if they support `shared_atomic` aspect
- A USM device buffer is accessible only from the device specified in the USM allocation.
```

1/ Clarify the current spec
	- Bullet Point / Section for accessibility
Optional: 
2/ Clarify the USM part (remove some sentences...)  [This can be useful for several reasons, such]	
3/ Sycl Next /  KHR  -> "Extentend accesibility": Another aspect? Not sure.
